### PR TITLE
Nerlworker

### PR DESCRIPTION
**Pull request by leondavi**

### CODE REVIEW

**Review by leondavi — COMMENTED** [review]
@Orisadek @GuyPerets106 
You have ruined something - we can't merge this! 
F1 Score is 0 it means that you messed up with basic perceptron! 
Please check the cause: 
I can blame dataset changes, train and predict functions in opennNNnif.cpp

File: src_cpp/opennnBridge/nerlWorkerOpenNN.cpp [hunk 1]
```diff
@@ -159,28 +423,30 @@ namespace nerlnet
                        LogError("NerlWorkerOpenNN::generate_custom_model_nn - PERCEPTRON cannot be first layer");
                        throw std::invalid_argument("NerlWorkerOpenNN::generate_custom_model_nn - PERCEPTRON cannot be first layer");
                     }
-                    std::shared_ptr<NerlLayer> prev_layer = curr_layer->get_prev_layer_ptr();
-                    int prev_layer_size = prev_layer->get_dim_size(DIM_X_IDX);
+                    int prev_layer_size;
+                    if(neural_network_ptr->get_layer_pointer(neural_network_ptr->get_layers_number()-1)->get_type_string() == "Flatten"){
```
Comment (leondavi) [inline src_cpp/opennnBridge/nerlWorkerOpenNN.cpp:427]: @Orisadek 
You must check if this code somehow influences the full flow test

File: src_cpp/opennnBridge/nerlWorkerOpenNN.cpp [hunk 2]
```diff
@@ -16,30 +17,119 @@ namespace nerlnet
         generate_opennn_neural_network();
         _training_strategy_ptr = std::make_shared<opennn::TrainingStrategy>();
         generate_training_strategy();
-        //TODO Ori and Nadav - implement training strategy (loss method, optimizer type, epochs, only sgd and adam)
+        _ae_red_ptr = std::make_shared<AeRed>();
```
Comment (leondavi) [inline src_cpp/opennnBridge/nerlWorkerOpenNN.cpp:20]: Add case only for AEC

File: src_cpp/opennnBridge/nerlWorkerOpenNN.cpp [hunk 3]
```diff
@@ -16,30 +17,119 @@ namespace nerlnet
...
+            case MODEL_TYPE_AUTOENCODER:
+            {
+                break;
+            }
+            case MODEL_TYPE_AE_CLASSIFIER:
+            {
+                std::shared_ptr<opennn::NeuralNetwork> neural_network = get_neural_network_ptr();
+                Index num_of_samples = _aec_data_set->dimension(0);
+                Index inputs_number = neural_network->get_inputs_number();
+                fTensor2DPtr calculate_res = std::make_shared<fTensor2D>(num_of_samples, neural_network->get_outputs_number());
```
Comment (leondavi) [inline src_cpp/opennnBridge/nerlWorkerOpenNN.cpp:94]: @GuyPerets106 
Do I miss something?!? 
calculate_res is an empty tensor; you don't even use the result_ptr which support to be the predicted labels tensor. 
What is going on here?

File: src_cpp/opennnBridge/openNNnif.cpp [hunk 4]
```diff
@@ -85,15 +78,23 @@ void* PredictFun(void* arg)
...
     Index inputs_number = neural_network->get_inputs_number();
-
     fTensor2DPtr calculate_res = std::make_shared<fTensor2D>(num_of_samples, neural_network->get_outputs_number());
+    Tensor<Index, 1> input_variable_dimension(4);
     Tensor<Index, 1> inputs_dimensions(2);
 
-    inputs_dimensions.setValues({num_of_samples, inputs_number});
-
-    *calculate_res = neural_network->calculate_outputs(PredictNNptr->data->data(), inputs_dimensions);
+    if(neural_network->has_convolutional_layer())
```
[the remark this answers is not in the document]
Comment (Orisadek) [reply src_cpp/opennnBridge/openNNnif.cpp:85]: > @Orisadek Why this code goes here? what its purpose? There is post_predict and pre_predict for this

In "calculate outputs" the function change according to the rank of the input.
in CNN we need to enter rank == 4  to get result.
neural_network->has_convolutional_layer() check if there is convolutional layer in the NN
I don't know what is pre-predict, but I saw the post predict.

File: src_cpp/opennnBridge/nerlWorkerOpenNN.cpp (hunk 3 above)
Comment (GuyPerets106) [reply src_cpp/opennnBridge/nerlWorkerOpenNN.cpp]: Fixed

**Pull request merged**